Python/traceback.c

```
@@ -393,12 +393,12 @@ _Py_WriteIndent(int indent, PyObject *f)
             buf[indent] = '\0';
         }
         err = PyFile_WriteString(buf, f);
-        if (err != 0) {
-            return err;
-        }
+        if (err == -1)
+            goto done;
         indent -= 10;
     }
-    return 0;
+done:
+    return err;
```

Did you consider using this "CPython idiom" throughout the PR (ditching `err`, where possible)?

```
    if (fn() < 0) {
        goto error;
    }

    return 0;

error:
    return -1;
}
```

IMHO, the advantage is that you've got two clear exit paths; I don't have to backtrack `err` to find out if the function will return 0 or -1 for each bad/good path. Probably just personal preference :)

---

In this case you're right, but there are places where there is some XDECREF that needs to happen regardless of the return value.

---

I guess it could be 

 ```
   Py_XDECREF(thing);
   return 0;

error:
    Py_XDECREF(thing);
    return -1;
}
```

---

True, in some situations, it may not be the best alternative because of ref counting; if you need a lot of decrefs before each goto, it is probably not worth it.

---

Or something like this:

```
    int ret = fn(obj);
    Py_DECREF(obj);
    if (ret < 0) {
        goto error;
    }

    return 0;

error:
    return -1;
```

---

So I did this everywhere and I agree it's much better now.

The only place where things are a little weird is tb_displayline, because there are cases where there is an early exit with return value 0 (depending on the return value of some "should I ignore this error" function).